crates/runtime/src/accelerated_table/timestamp_metrics_utils.rs

```diff
@@ -0,0 +1,655 @@
+use crate::component::dataset::TimeFormat;
+use crate::dataupdate::StreamingDataUpdate;
+use arrow::array::{
+    Array, Date32Array, Float16Array, Float32Array, Float64Array, Int8Array, Int16Array,
+    Int32Array, Int64Array, LargeStringArray, StringArray, TimestampMicrosecondArray,
+    TimestampMillisecondArray, TimestampNanosecondArray, TimestampSecondArray, UInt8Array,
+    UInt16Array, UInt32Array, UInt64Array,
+};
+use arrow::datatypes::TimeUnit;
+use arrow_schema::{DataType, SchemaRef};
+use async_stream::stream;
+use chrono::DateTime;
+use datafusion::execution::SendableRecordBatchStream;
+use datafusion::physical_plan::stream::RecordBatchStreamAdapter;
+use futures::StreamExt;
+use std::sync::Arc;
+use tokio::sync::Mutex;
+
+const MS_IN_A_DAY: i64 = 24 * 60 * 60 * 1000;
+
+macro_rules! max_ts_macro {
+    ($ty:ty, $max_ts:expr, $array:expr, $value:expr) => {{
+        match $array.as_any().downcast_ref::<$ty>() {
+            Some(arr) => {
+                let batch_max = (0..arr.len())
+                    .filter(|&i| !arr.is_null(i))
+                    .filter_map(|i| $value(arr, i))
+                    .max();
+                match batch_max {
+                    Some(bm) => Some($max_ts.map_or(bm, |cur| cur.max(bm))),
+                    None => $max_ts,
+                }
+            }
+            None => $max_ts,
+        }
+    }};
+}
+
+/// Extracts the maximum timestamp from a stream of record batches.
+///
+/// The extraction uses `time_column` and `time_format` to interpret timestamps
+/// from various data types (timestamps, strings, integers, floats, dates).
+///
+/// # Parameters
+/// - `time_column`: Name of the column containing timestamp data
+/// - `time_format`: Format specification for interpreting the timestamp values
+/// - `source_name`: Used for logging warnings
+///
+/// # Returns
+/// - An updated data stream
+/// - An `Option<Arc<Mutex<Option<i64>>>>` with the following semantics:
+///     - `None`: Did not attempt to find a max timestamp (e.g., no `time_column`
+///       provided or column does not exist in schema)
+///     - `Some(Arc<Mutex<Option<i64>>>)`: Will be populated after the stream is consumed:
+///         - `Some(i64)`: The maximum timestamp found in the stream (in milliseconds)
+///         - `None`: Attempted extraction but no valid timestamps found
+///             - If batches were not empty, a warning is logged
+#[allow(clippy::too_many_lines)]
+pub async fn max_timestamp_in_stream(
```

Comment:
Nit: another alternative to receive max_timestamp is to use oneshot channel, this is JFYI, I don't think there is any performance difference.